### ISSUE TITLE
Problems reproducing LDM for conditional imageNet

### ISSUE THREAD

**Issue by neudachina** [vinhsuhi/LD3#3]
Dear Authors,

We have successfully reproduced the result you have reported on pixel datasets and LSUN dataset for latent model. But we came up with the problem achieving FID 4.32 on conditional ImageNet. Your code results in FID=5.08. Changing your uni-pc solver variant from bh2 to bh1 makes it better, but still not close enough (LSUN dataset was reproduced very closely, up to three decimals). We use the model checkpoint and FID reference statistics downloaded from links in download_models.sh script

We have also noticed that you come uses "random" class sampling, while your paper reports the use of uniform sampling of ImageNet classes

Could you please help resolve these issues, or is there any bugs in the current settings for conditional LDM model?
Thanks a lot

**Comment by vinhsuhi** [comment]
Thanks for your interest in our work.

Regarding the conditional ImageNet results: for training, we use random class sampling, but for FID evaluation, we employ uniform sampling across classes to ensure a balanced distribution. This detail was unfortunately overlooked in the released code—thank you for pointing it out. Your observation is absolutely valid.

Have you tried setting "uniform" instead of "random" ([https://github.com/vinhsuhi/LD3/blob/main/gen_data.py#L93](https://github.com/vinhsuhi/LD3/blob/main/gen_data.py#L93))? This should help align your evaluation setup with ours and potentially reduce the FID gap.

Let us know if you have further questions. We’re happy to help.

**Comment by neudachina** [comment]
Thank you for such a quick answer for our issue

Using the code you have released and changing random sampling to uniform results in FID = 5.062 with default seed=0. I used batch_size=50 because I noticed that conditioning sampling depends on it

It is also worth noticing, that I use an environment created with your requirements.yaml file. so that the problem couldn't be connected to this. FID is calculated with EDM code, it works great with LSUN

Do you have any ideas, why do we have trouble reproducing your results with this dataset?

Thank you for your further answer and for being open to helping us

**Comment by hoangdzung** [comment]
Hi @neudachina,

For latent-diffusion models, we recommend using the FID evaluation code from [`pytorch_fid`](https://github.com/mseitzer/pytorch-fid), which relies on this version of [INCEPTION-V3 model](https://github.com/mseitzer/pytorch-fid/blob/master/src/pytorch_fid/inception.py#L13C1-L14C1). Different evaluation pipelines can produce varying FID values.

Please try `pytorch_fid` and let us know if you need further help.

**Comment by neudachina** [comment]
I have just tried to calculate FID using pytorch-fid==0.3.0 (the one you have denoted in requirements.yaml) and still got 5.061 :((

**Comment by neudachina** [comment]
To better clarify our issue, there are more details provided:

- we are using your script to generate images:
```
python3 gen_data.py \
--all_config configs/latent_diff_imn.yml \
--total_samples 50000 \
--sampling_batch_size 50 \
--steps 10 \
--solver_name uni_pc \
--skip_type time_uniform \
--save_png --data_dir train_data_test/train_data_imn
``` 

- the only thing we change in you code is swapping random with uniform conditional sampling in `gen_data.py` file

- to calculate FID `python -m pytorch_fid train_data_test/train_data_imn/uni_pc_NFE10_time_uniform_seed0 fid-refs/VIRTUAL_imagenet256_labeled.npz` is used
 
- FID refs and model checkpoints are downloaded via `download_models.sh` script

Is there anything we are missing in generating teacher dataset for your method?

**Comment by hoangdzung** [comment]
Thanks for the information.

Actually, the teacher for latent diffusion ImageNet uses `time_quadratic` with 10 steps, not `time_uniform`. It’s a typo in our paper. Thank you for catching this. We will correct this in the final camera-ready version.

Let us know if you can now reproduce the numbers or if you encounter any other issues.

**Comment by neudachina** [comment]
Thanks for you answer!

Unfortunately, using `time_quadratic` and `cfg=2.0` results in FID=4.17. It is the closest we have been to the numbers reported in your paper but still not close enough

Additionally, we have a question about `cfg` scale. You reported using `cgf=2.0` but we've noted that LDM paper authors use `cfg=1.5` saying that it works better. In our experiments we found out that:

`time_quadratic` with `cfg=2.0` results in `4.17`
`time_quadratic` with `cfg=1.5` results in `4.98`

`time_uniform` with `cfg=2.0` results in `5.043`
`time_uniform` with `cfg=1.5` results in `3.850`

We would like to clarify with you again the configuration you are using. Is there any way you have used a classifier-free guidance scale different to `2.0`?

**Comment by vinhsuhi** [comment]
Thanks for your detailed follow-up!

You're absolutely right — in our paper, we did not fix the random seed when reporting results, so there might be some variance in the numbers. However, based on our experience, your FID scores seem to be within a reasonable range of what we observed.

Regarding the CFG scale: yes, we consistently used `cfg=2.0` in all our experiments. We are aware that the LDM paper suggests `cfg=1.5` for better results in some cases, and it’s interesting to see that your results reflect this behavior for the `time_uniform` schedule. That said, we did not experiment with other CFG values extensively, so your observations are definitely valuable and might help others tuning their setups.

Let us know if you'd like to compare more results or need help with specific configurations!

**Comment by neudachina** [comment]
Thank you so much for your cooperation and for taking the time to address this issue!
